File: flutter/nativebrik_bridge/ios/Classes/Manager.swift
//
//  Manager.swift
//  nativebrik_bridge
//
//  Created by Ryosuke Suzuki on 2024/03/10.
//

import Foundation
import Flutter
import UIKit
import Nativebrik

struct EmbeddingEntity {
    let uiview: UIView
    let channel: FlutterMethodChannel
    let accessor: __DO_NOT_USE__NativebrikBridgedViewAccessor?
}

struct RemoteConfigEntity {
    let variant: RemoteConfigVariant?
}

class NativebrikBridgeManager {
    private var nativebrikClient: NativebrikClient? = nil
    private var embeddingMaps: [String:EmbeddingEntity]
    private var configMaps: [String:RemoteConfigEntity]

    init() {
        self.embeddingMaps = [:]
        self.configMaps = [:]
    }

    func setNativebrikClient(nativebrik: NativebrikClient) {
        if self.nativebrikClient != nil {
            return print("NativebrikClient is already set")
        }
        self.nativebrikClient = nativebrik
        if let vc = UIApplication.shared.delegate?.window??.rootViewController {
            let overlay = nativebrik.experiment.overlayViewController()
            vc.addChild(overlay)
            vc.view.addSubview(overlay.view)
        }
    }

    func getUserId() -> String? {
        guard let nativebrikClient = self.nativebrikClient else {
            return nil
        }
        return nativebrikClient.user.id
    }

    func setUserProperties(properties: [String: String]) {
        guard let nativebrikClient = self.nativebrikClient else {
            return
        }
        nativebrikClient.user.setProperties(properties)
    }

    func getUserProperties() -> [String: String]? {
        guard let nativebrikClient = self.nativebrikClient else {
            return nil
        }
        return nativebrikClient.user.getProperties()
    }

    // embedding
    func connectEmbedding(id: String, channelId: String, arguments: Any?, messenger: FlutterBinaryMessenger) {
        guard let nativebrikClient = self.nativebrikClient else {
            return
        }
        let channel = FlutterMethodChannel(name: "Nativebrik/Embedding/\(channelId)", binaryMessenger: messenger)
        let uiview = nativebrikClient.experiment.embeddingUIView(id, arguments: arguments, onEvent: { event in
            channel.invokeMethod(ON_EVENT_METHOD, arguments: [
                "name": event.name as Any?,
                "deepLink": event.deepLink as Any?,
                "payload": event.payload?.map({ prop in
                    return [
                        "name": prop.name,
                        "value": prop.value,
                        "type": prop.type
                    ]
                }),
            ])
        }) { phase in
            switch phase {
            case .completed(let view):
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "completed")
                return view
            case .notFound:
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "not-found")
                return UIView()
            case .failed:
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "failed")
                return UIView()
            case .loading:
                return UIView()
            }
        }
        let embeedingEntity = EmbeddingEntity(
            uiview: uiview,
            channel: channel,
            accessor: nil
        )
        self.embeddingMaps[channelId] = embeedingEntity
    }

    func disconnectEmbedding(channelId: String) {
        self.embeddingMaps[channelId] = nil
    }

    func getEmbeddingEntity(channelId: String) -> EmbeddingEntity? {
        guard let entity = self.embeddingMaps[channelId] else {
            return nil
        }
        print("get embedding entity \(channelId)")
        return entity
    }

    // remote config
    func connectRemoteConfig(id: String, channelId: String, onPhase:  @escaping ((RemoteConfigPhase) -> Void)) {
        guard let nativebrikClient = self.nativebrikClient else {
            return
        }
        let entity = RemoteConfigEntity(variant: nil)
        self.configMaps[channelId] = entity

        nativebrikClient.experiment.remoteConfig(id) { phase in
            switch phase {
            case .completed(let config):
                if self.configMaps[channelId] == nil {
                    // disconnected already
                    return
                }
                let entity = RemoteConfigEntity(variant: config)
                self.configMaps[channelId] = entity
                onPhase(phase)
            case .notFound:
                onPhase(phase)
            case .failed:
                onPhase(phase)
            default:
                break
            }
        }
    }

    func disconnectRemoteConfig(channelId: String) {
        self.configMaps[channelId] = nil
    }

    func connectEmbeddingInRemoteConfigValue(key: String, channelId: String, arguments: Any?, embeddingChannelId: String, messenger: FlutterBinaryMessenger) {
        guard let entity = self.configMaps[channelId] else {
            return
        }
        guard let variant = entity.variant else {
            return
        }
        let channel = FlutterMethodChannel(name: "Nativebrik/Embedding/\(embeddingChannelId)", binaryMessenger: messenger)
        guard let uiview = variant.getAsUIView(key, arguments: arguments, onEvent: { event in
            channel.invokeMethod(ON_EVENT_METHOD, arguments: [
                "name": event.name as Any?,
                "deepLink": event.deepLink as Any?,
                "payload": event.payload?.map({ prop in
                    return [
                        "name": prop.name,
                        "value": prop.value,
                        "type": prop.type
                    ]
                }),
            ])
        }, content: { phase in
            switch phase {
            case .completed(let view):
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "completed")
                return view
            case .notFound:
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "not-found")
                return UIView()
            case .failed:
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "failed")
                return UIView()
            case .loading:
                return UIView()
            }
        }) else {
            return
        }
        let embeedingEntity = EmbeddingEntity(
            uiview: uiview,
            channel: channel,
            accessor: nil
        )
        self.embeddingMaps[embeddingChannelId] = embeedingEntity
    }

    func getRemoteConfigValue(channelId: String, key: String) -> String? {
        guard let entity = self.configMaps[channelId] else {
            return nil
        }
        guard let variant = entity.variant else {
            return nil
        }
        return variant.get(key)
    }

    // tooltip
    func connectTooltip(name: String, onFetch: @escaping (String) -> Void, onError: @escaping (String) -> Void) {
        guard let nativebrikClient = self.nativebrikClient else {
            onError("NativebrikClient is not set")
            return
        }
        Task {
            let result = await nativebrikClient.experiment.__do_not_use__fetch_tooltip_data(trigger: name)
            switch result {
            case .success(let data):
                onFetch(data)
            case .failure(let error):
                onError(error.localizedDescription)
            }
        }
    }

    func connectTooltipEmbedding(channelId: String, rootBlock: String, messenger: FlutterBinaryMessenger) {
        guard let nativebrikClient = self.nativebrikClient else {
            return
        }
        let channel = FlutterMethodChannel(name: "Nativebrik/Embedding/\(channelId)", binaryMessenger: messenger)
        print("connect tooltip embedding", channelId)
        let accessor = nativebrikClient.experiment.__do_not_use__render_uiview(
            json: rootBlock,
            onEvent: { event in
                channel.invokeMethod(ON_EVENT_METHOD, arguments: [
                    "name": event.name as Any?,
                    "deepLink": event.deepLink as Any?,
                    "payload": event.payload?.map({ prop in
                        return [
                            "name": prop.name,
                            "value": prop.value,
                            "type": prop.type
                        ]
                    }),
                ])
            },
            onNextTooltip: { pageId in
                print("NativebrikBridgeManager onNextTooltip: \(pageId))")
                channel.invokeMethod(ON_NEXT_TOOLTIP_METHOD, arguments: [
                    "pageId": pageId,
                ])
            },
            onDismiss: {
                channel.invokeMethod(ON_DISMISS_TOOLTIP_METHOD, arguments: nil)
            }
        )
        print("accessor", accessor)
        let embeedingEntity = EmbeddingEntity(
            uiview: accessor.view,
            channel: channel,
            accessor: accessor
        )
        print("add emedding: \(channelId)")
        self.embeddingMaps[channelId] = embeedingEntity
    }
    
    func callTooltipEmbeddingDispatch(channelId: String, event: String) {
        guard let entity = self.embeddingMaps[channelId] else {
            return
        }
        guard let accessor = entity.accessor else {
            return
        }
        do {
            try accessor.dispatch(event: event)
        } catch {}
    }

    func disconnectTooltipEmbedding(channelId: String) {
        self.embeddingMaps[channelId] = nil
    }

    // trigger
    func dispatch(name: String) {
        guard let nativebrikClient = self.nativebrikClient else {
            return
        }
        nativebrikClient.experiment.dispatch(NativebrikEvent(name))
    }

    /**
     * Records a crash with the given exception.
     *
     * This method forwards the exception to the Nativebrik SDK for crash reporting.
     */
    func recordCrash(exception: NSException) {
        guard let nativebrikClient = self.nativebrikClient else {
            return
        }
        nativebrikClient.experiment.record(exception: exception)
    }
}

File: flutter/nativebrik_bridge/android/src/main/kotlin/com/nativebrik/flutter/nativebrik_bridge/Manager.kt
package com.nativebrik.flutter.nativebrik_bridge

import android.content.Context
import com.nativebrik.sdk.NativebrikClient
import com.nativebrik.sdk.__DO_NOT_USE_THIS_INTERNAL_BRIDGE
import com.nativebrik.sdk.data.NotFoundException
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import io.flutter.plugin.common.StandardMessageCodec
import android.view.View
import android.widget.LinearLayout
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import com.nativebrik.sdk.NativebrikEvent
import com.nativebrik.sdk.remoteconfig.RemoteConfigVariant
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

internal data class ConfigEntity(val variant: RemoteConfigVariant?, val experimentId: String?)

internal class NativebrikBridgeManager(private val binaryMessenger: BinaryMessenger) {
    private var nativebrikClient: NativebrikClient? = null
    private var bridgeClient: __DO_NOT_USE_THIS_INTERNAL_BRIDGE? = null

    private var embeddingMap: MutableMap<String, Any?> = mutableMapOf()
    private var configMap: MutableMap<String, ConfigEntity> = mutableMapOf()

    fun setNativebrikClient(client: NativebrikClient) {
        this.nativebrikClient = client
        this.bridgeClient = __DO_NOT_USE_THIS_INTERNAL_BRIDGE(client)
    }

    fun getUserId(): String? {
        return this.nativebrikClient?.user.id
    }

    fun setUserProperties(properties: Map<String, String>) {
        this.nativebrikClient?.user.setProperties(properties)
    }

    fun getUserProperties(): Map<String, String>? {
        return this.nativebrikClient?.user.getProperties()
    }

    // embedding
    @OptIn(DelicateCoroutinesApi::class)
    fun connectEmbedding(channelId: String, experimentId: String, componentId: String? = null) {
        val methodChannel = MethodChannel(this.binaryMessenger, "Nativebrik/Embedding/$channelId")
        GlobalScope.launch(Dispatchers.IO) {
            val result = bridgeClient?.connectEmbedding(experimentId, componentId)
            if (result == null) {
                GlobalScope.launch(Dispatchers.Main) {
                    methodChannel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, "not-found")
                }
                return@launch
            }
            result.onSuccess {
                embeddingMap[channelId] = it
                GlobalScope.launch(Dispatchers.Main) {
                    methodChannel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, "completed")
                }
            }.onFailure {
                when (it) {
                    is NotFoundException -> {
                        GlobalScope.launch(Dispatchers.Main) {
                            methodChannel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, "not-found")
                        }
                    }
                    else -> {
                        GlobalScope.launch(Dispatchers.Main) {
                            methodChannel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, "failed")
                        }
                    }
                }
            }
        }
    }

    fun disconnectEmbedding(channelId: String) {
        embeddingMap.remove(channelId)
    }

    @Composable
    fun Render(channelId: String, arguments: Any?, modifier: Modifier = Modifier) {
        if (channelId.isEmpty()) {
            return
        }
        val bridgeClient = this.bridgeClient ?: return
        val methodChannel = remember(channelId) {
            MethodChannel(this.binaryMessenger, "Nativebrik/Embedding/$channelId")
        }
        val data = this.embeddingMap[channelId]
        bridgeClient.render(modifier, arguments, data, onEvent = { event ->
            methodChannel.invokeMethod(ON_EVENT_METHOD, mapOf(
                "name" to event.name,
                "deepLink" to event.deepLink,
                "payload" to event.payload?.map { prop ->
                    mapOf(
                        "name" to prop.name,
                        "value" to prop.value,
                        "type" to prop.type,
                    )
                }
            ))
        })
    }

    @Composable
    fun RenderOverlay() {
        nativebrikClient?.experiment?.Overlay()
    }

    // remote config
    suspend fun connectRemoteConfig(channelId: String, experimentId: String): Result<String> {
        if (channelId.isEmpty()) {
            return Result.success("not-found")
        }
        this.configMap[channelId] = ConfigEntity(null, null)

        if (experimentId.isEmpty()) {
            return Result.success("not-found")
        }
        val client = this.nativebrikClient ?: return Result.success("not-found")
        val config = client.experiment.remoteConfig(experimentId)
        val variant = config.fetch().getOrElse {
            val status = when (it) {
                is NotFoundException -> "not-found"
                else -> "failed"
            }
            return Result.success(status)
        }
        if (this.configMap[channelId] != null) {
            this.configMap[channelId] = ConfigEntity(variant, variant.experimentId)
        }
        return Result.success("competed")
    }

    fun disconnectRemoteConfig(channelId: String) {
        this.configMap.remove(channelId)
    }

    fun getRemoteConfigValue(channelId: String, key: String): String? {
        if (channelId.isEmpty()) return null
        if (key.isEmpty()) return null
        val config = this.configMap[channelId] ?: return null
        val variant = config.variant ?: return null
        return variant.get(key)
    }

    fun connectEmbeddingInRemoteConfigValue(channelId: String, key: String, embeddingChannelId: String) {
        if (channelId.isEmpty()) return
        val config = this.configMap[channelId] ?: return
        val variant = config.variant ?: return
        val componentId = variant.get(key) ?: return
        val experimentId = config.experimentId ?: return
        this.connectEmbedding(embeddingChannelId, experimentId, componentId)
    }

    fun dispatch(name: String) {
        this.nativebrikClient?.experiment?.dispatch(NativebrikEvent(name))
    }
}

internal class OverlayViewFactory(private val manager: NativebrikBridgeManager): PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        return OverlayView(context, manager)
    }
}

internal class OverlayView(context: Context, manager: NativebrikBridgeManager): PlatformView {
    private val view: ComposeView

    override fun getView(): View {
        return view
    }

    override fun dispose() {}

    init {
        view = ComposeView(context).apply {
            setContent {
                manager.RenderOverlay()
            }
        }
    }
}

internal class NativeViewFactory(private val manager: NativebrikBridgeManager): PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val creationParams = args as Map<*, *>?
        val channelId = creationParams?.get("channelId") as String
        val arguments = creationParams["arguments"]
        return NativeView(context, channelId, arguments, manager)
    }
}

internal class NativeView(context: Context, channelId: String, arguments: Any?, manager: NativebrikBridgeManager): PlatformView {
    private val view: ComposeView

    override fun getView(): View {
        return view
    }

    override fun dispose() {}

    init {
        val param = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT)
        view = ComposeView(context).apply {
            setContent {
                manager.Render(channelId, arguments)
            }
            layoutParams = param
        }
    }
}